### PR TITLE
[FR] Add IPv6 Support to CidrMatch using ipaddress lib

### PR DESCRIPTION
**Pull request by eric-forte-elastic**
## Summary

Relates https://github.com/elastic/ia-trade-team/issues/196

Alternative to https://github.com/endgameinc/eql/pull/77 where the ipaddres library is used instead. 

## Testing

The majority of the testing can be performed via the new `ipv6` labeled unit tests in `test_python_engine.py` and `test_cidrmatch` in `test_functions.py`. However, additional testing can be performed by calling the functions directly from a python script. 

<details><summary>Example Python Script</summary>
<p>

```python
from eql.functions import CidrMatch

# create a list of CIDR blocks to match against
cidr_blocks = ["192.168.0.0/24", "2001:db8::/32"]

# create an instance of the CidrMatch class
cidr_matcher = CidrMatch()


# call the run method with an IP address and the list of CIDR blocks
ip_address = "192.168.0.1"
result = cidr_matcher.run(ip_address, *cidr_blocks)

# print the result
print(result)  # True


# create a list of CIDR blocks to match against
cidr_blocks = ["2001:0db8:0000:0000:0000:0000:0000:0000/32", "fe80::/10"]

# create an instance of the CidrMatch class
cidr_matcher = CidrMatch()

# call the run method with an IPv6 address and the list of CIDR blocks
ipv6_address = "2001:0db8:0000:0000:0000:0000:0000:0001"
ipv6_address = "2001:db8::1"
ipv6_address = "fe80::1"
print(ipv6_address)
result = cidr_matcher.run(ipv6_address, *cidr_blocks)
print(result)

ipv6_address = CidrMatch.expand_ipv6_address(ipv6_address)
print(ipv6_address)
result = cidr_matcher.run(ipv6_address, *cidr_blocks)

# print the result
print(result)  # True

```

</p>
</details> 

:red_circle: **Important** :red_circle: 

The testing MUST be performed on both Python 3.7 and Python 2.7 as eql needs to be able to support both. Some editors such VSCode do not support debugging in Python 2.7 so one will have to run the unit tests via another method. 

To install eql in a Python2 test environment you can follow the following steps (there are other ways to test too, but this can help get started)
1. Build virtualenv: `virtualenv --python=python2.7 test_two_venv`
1. Activate virtualenv: `source /test_two_venv/bin/activate`
1. Install setuptools (as Makefile does as well) `pip install setuptools -U`
1. [Optional] if you have a py3 pypirc, you may need to move it temporarily e.g. `mv /home/forteea1/.pypirc ~/pypirc`
1. Install new eql version (check that it says 0.9.19) `python setup.py install`
1. Run your test script (example above) `python test.py`

### CODE REVIEW

File: eql/utils.py [hunk 1]
```diff
@@ -79,6 +83,20 @@ def str_presenter(dumper, data):
...
+    if py_version == 2:
+        ipaddr_string = ipaddr_string.decode("utf-8")  # noqa: F821
+    return ipaddress.ip_address(ipaddr_string)
+
+
+def get_subnet(cidr_string):
+    """Get an ipaddress ip_network object from a string containing an cidr range."""
+    if py_version == 2:
+        cidr_string = cidr_string.decode("utf-8")  # noqa: F821
+    return ipaddress.ip_network(cidr_string, strict=False)
```
Comment (Mikaayenson) [inline eql/utils.py:97]: Can you add type hints here?

File: eql/utils.py [hunk 2]
```diff
@@ -79,6 +83,20 @@ def str_presenter(dumper, data):
...
 
 
+def get_ipaddress(ipaddr_string):
+    """Get an ipaddress ip_address object from a string containing an ipaddress."""
+    if py_version == 2:
+        ipaddr_string = ipaddr_string.decode("utf-8")  # noqa: F821
+    return ipaddress.ip_address(ipaddr_string)
+
+
+def get_subnet(cidr_string):
```
Comment (Mikaayenson) [inline eql/utils.py:93]: It might be good to pass in `strict` param vs defaulting.

File: eql/functions.py [hunk 3]
```diff
@@ -349,14 +249,13 @@ def validate(cls, arguments):
             # overwrite the original node
             text = argument.node.value.strip()
 
-            if not cls.cidr_compiled.match(argument.node.value):
+            if not cls.is_cidr(text):
```
Comment (Mikaayenson) [inline eql/functions.py:252]: Do we want this to default to `strict=false` when calling `get_subnet`?

File: eql/functions.py [hunk 4]
```diff
@@ -322,17 +210,29 @@ def callback(source, *_):
...
+                    if ip_address in subnet:
                         return True
 
         return False
 
+    @classmethod
+    def is_cidr(cls, cidr):
+        """Check if a string is a valid CIDR notation."""
+        if "/" not in cidr:
+            return False
```
Comment (Mikaayenson) [inline eql/functions.py:229]: Since were calling `get_subnet` do we even need this slash check?

**Review by Mikaayenson — APPROVED** [review]
A few small nits, but LGTM.

File: eql/utils.py [hunk 5]
```diff
@@ -79,6 +83,20 @@ def str_presenter(dumper, data):
     return dumper.represent_scalar('tag:yaml.org,2002:str', data)
 
 
+def get_ipaddress(ipaddr_string):
+    """Get an ipaddress ip_address object from a string containing an ipaddress."""
```
Comment (brokensound77) [inline eql/utils.py:87]:
```suggestion
    """Get an ip_address object from a string containing an ip address."""
```

File: eql/utils.py [hunk 6]
```diff
@@ -79,6 +83,20 @@ def str_presenter(dumper, data):
...
+    if py_version == 2:
+        ipaddr_string = ipaddr_string.decode("utf-8")  # noqa: F821
+    return ipaddress.ip_address(ipaddr_string)
+
+
+def get_subnet(cidr_string, strict=False):
+    """Get an ipaddress ip_network object from a string containing an cidr range."""
```
Comment (brokensound77) [inline eql/utils.py:94]:
```suggestion
    """Get an ip_network object from a string containing an cidr range."""
```

File: tests/test_python_engine.py [hunk 7]
```diff
@@ -389,6 +392,338 @@ def test_custom_functions(self):
...
+        """
+        with elasticsearch_syntax:
+            parsed_query = parse_query(query)
+
+        output = self.get_output(queries=[parsed_query], config=config, events=events)
+
+        self.assertEqual(len(output), 0, "Missing or extra results")
+
+        # Should return results since the cidr ranges match the addresses in the events
+        query = """
```
Comment (brokensound77) [inline tests/test_python_engine.py:494]: Can you add a test that searches the known `events` where the shorthand is explicitly defined in the query (to validate expansion)

File: eql/functions.py [hunk 8]
```diff
@@ -322,17 +210,29 @@ def callback(source, *_):
...
+                if is_string(cidr):
+                    subnet = get_subnet(cidr)
+
+                    if ip_address in subnet:
                         return True
 
         return False
 
+    @classmethod
+    def is_cidr(cls, cidr):
```
Comment (brokensound77) [inline eql/functions.py:226]:
```suggestion
    def is_cidr_pattern(cls, cidr):
```

or `is_cidr_notation`

we can also move this to utils as well I would think

**Review by brokensound77 — APPROVED** [review]
A few comments otherwise, LGTM 👍

**Review by terrancedejesus — APPROVED** [review]

**Pull request merged**